rust/lance-linalg/src/distance/l2.rs

```diff
@@ -93,43 +130,52 @@ impl L2 for [f32] {
             let mut sum = f32x8::zeros();
             for i in (0..len).step_by(8) {
                 unsafe {
-                    let mut x = f32x8::load_unaligned(self.as_ptr().add(i));
-                    let y = f32x8::load_unaligned(other.as_ptr().add(i));
+                    let mut x = f32x8::load_unaligned(x.as_ptr().add(i));
+                    let y = f32x8::load_unaligned(y.as_ptr().add(i));
                     x -= y;
                     sum.multiply_add(x, x);
                 }
             }
             sum.reduce_sum()
         } else {
             // Fallback to scalar
-            l2_scalar(self, other)
+            l2_scalar::<f32, 16>(x, y)
         }
     }
-}
-
-impl L2 for Float32Array {
-    type Output = f32;
 
-    #[inline]
-    fn l2(&self, other: &Self) -> f32 {
-        self.values().l2(other.values())
+    fn l2_batch<'a>(
+        x: &'a [Self::Native],
+        y: &'a [Self::Native],
+        dimension: usize,
+    ) -> Box<dyn Iterator<Item = Self::Native> + 'a> {
+        use self::f32::l2_once;
+        // Dispatch based on the dimension.
+        match dimension {
+            8 => Box::new(
+                y.chunks_exact(dimension)
+                    .map(move |v| l2_once::<f32x8, 8>(x, v)),
+            ),
+            16 => Box::new(
+                y.chunks_exact(dimension)
+                    .map(move |v| l2_once::<f32x16, 16>(x, v)),
+            ),
+            _ => Box::new(y.chunks_exact(dimension).map(|v| Self::l2(x, v))),
```

Comment:
Will this always resolve correctly?

Reply:
We let the caller to make sure y is a batch of centroids. Even the dimension is odd number, `y.len() % dimension == 0`.  We could prob add a `debug_assert` here.